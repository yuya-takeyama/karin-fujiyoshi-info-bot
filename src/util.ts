import { APIGatewayProxyResult } from 'aws-lambda';
import { Feed } from 'feed';

const baseUrl = 'http://www.keyakizaka46.com';
export const path = (path: string) => baseUrl + path;

export const ensureError = (err: any): Error => {
  if (err instanceof Error) {
    return err;
  } else if (err && typeof err.message === 'string') {
    return new Error(err.message);
  } else {
    return new Error(err);
  }
};

export const feedToResponse = (feed: Feed): APIGatewayProxyResult => {
  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'application/atom+xml',
    },
    body: feed.atom1(),
  };
};

export const errorToResponse = (error: Error): APIGatewayProxyResult => {
  return {
    statusCode: 500,
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      message: error.message,
    }),
  };
};
