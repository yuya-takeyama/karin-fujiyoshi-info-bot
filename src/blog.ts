import { JSDOM } from 'jsdom';
import striptags from 'striptags';
import { Feed } from 'feed';
import { parseFromTimeZone } from 'date-fns-timezone';
import { path } from './util';
import fetch from 'node-fetch';

type BlogArticle = {
  title: string;
  url: string;
  date: Date;
};

type AtomOptions = {
  updated: Date;
};

export const fetchBlogPage = async (): Promise<string> => {
  const res = await fetch(path('/s/k46o/diary/member/list?ima=0000&ct=47'));

  return res.text();
};

export const blogHtmlToAtom = (html: string, options: AtomOptions): string => {
  const articles = blogHtmlToArticles(html);

  return articlesToAtom(articles, options);
};

const articlesToAtom = (
  articles: BlogArticle[],
  options: AtomOptions,
): string => {
  const feed = new Feed({
    title: '欅坂46 藤吉 夏鈴 公式ブログ',
    description: '「坂道シリーズ」第2弾　欅坂46',
    feed: path('/'),
    id: path('/'),
    link: path('/'),
    feedLinks: {},
    copyright: 'Seed & Flower LLC',
    updated: options.updated,
  });

  for (const article of articles) {
    feed.addItem({
      title: article.title,
      link: article.url,
      date: article.date,
    });
  }

  return feed.atom1();
};

const blogHtmlToArticles = (html: string) => {
  const dom = new JSDOM(html);
  const elems = dom.window.document.querySelectorAll(
    '.box-content .box-main article',
  );
  const articles: BlogArticle[] = [];
  for (const elem of elems) {
    const titleElem = elem.querySelector('.box-ttl h3 a');
    const dateElem = elem.querySelector('.box-bottom ul li');
    if (titleElem && dateElem) {
      const date = parseFromTimeZone(striptags(dateElem.innerHTML).trim(), {
        timeZone: 'Asia/Tokyo',
      });
      const url = titleElem.getAttribute('href');
      if (typeof url === 'string') {
        articles.push({
          title: striptags(titleElem.innerHTML).trim(),
          url: path(url),
          date: date,
        });
      }
    }
  }
  return articles;
};
