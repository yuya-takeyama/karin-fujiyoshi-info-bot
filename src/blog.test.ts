import { join } from 'path';
import { readFile as origReadFile } from 'fs';
import { promisify } from 'util';
import { blogHtmlToFeed } from './blog';

const readFile = promisify(origReadFile);

const blogFixturesDir = join(__dirname, 'fixtures', 'blog');

describe('#blogHtmlToFeed', () => {
  it('returns a Feed', async () => {
    const html = (
      await readFile(join(blogFixturesDir, 'blog.html'))
    ).toString();
    expect(blogHtmlToFeed(html)).toMatchSnapshot();
  });
});
