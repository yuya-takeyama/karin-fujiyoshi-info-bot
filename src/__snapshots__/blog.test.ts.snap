// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#blogHtmlToFeed returns a Feed 1`] = `
Feed {
  "addCategory": [Function],
  "addContributor": [Function],
  "addExtension": [Function],
  "addItem": [Function],
  "atom1": [Function],
  "categories": Array [],
  "contributors": Array [],
  "extensions": Array [],
  "items": Array [
    Object {
      "date": 2020-09-05T10:23:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/35467?ima=0000&cd=member",
      "title": "9",
    },
    Object {
      "date": 2020-08-29T08:32:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/35348?ima=0000&cd=member",
      "title": "19",
    },
    Object {
      "date": 2020-08-16T14:58:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/35181?ima=0000&cd=member",
      "title": "夏だ",
    },
    Object {
      "date": 2020-08-08T07:02:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/35075?ima=0000&cd=member",
      "title": "お久しぶりです",
    },
    Object {
      "date": 2020-07-19T13:19:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/34842?ima=0000&cd=member",
      "title": "愛おしい存在",
    },
    Object {
      "date": 2020-07-15T21:37:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/34788?ima=0000&cd=member",
      "title": "お久しぶりです",
    },
    Object {
      "date": 2020-06-12T07:58:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/34288?ima=0000&cd=member",
      "title": "さてさて",
    },
    Object {
      "date": 2020-05-29T08:21:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/34050?ima=0000&cd=member",
      "title": "うーん",
    },
    Object {
      "date": 2020-05-23T05:36:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/33949?ima=0000&cd=member",
      "title": "眠",
    },
    Object {
      "date": 2020-05-16T10:36:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/33827?ima=0000&cd=member",
      "title": "お",
    },
    Object {
      "date": 2020-05-10T14:37:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/33739?ima=0000&cd=member",
      "title": "猫",
    },
    Object {
      "date": 2020-05-03T11:53:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/33630?ima=0000&cd=member",
      "title": "とにかく春が苦手ナンデス",
    },
    Object {
      "date": 2020-04-15T14:13:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/33357?ima=0000&cd=member",
      "title": "これは",
    },
    Object {
      "date": 2020-04-11T11:43:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/33299?ima=0000&cd=member",
      "title": "おー",
    },
    Object {
      "date": 2020-04-06T12:54:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/33216?ima=0000&cd=member",
      "title": "4周年だーー",
    },
    Object {
      "date": 2020-04-05T11:47:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/33144?ima=0000&cd=member",
      "title": "いつまでもセーラー服が着たいんだ!!!!!",
    },
    Object {
      "date": 2020-04-04T08:15:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/33126?ima=0000&cd=member",
      "title": "4月",
    },
    Object {
      "date": 2020-03-22T14:29:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/32995?ima=0000&cd=member",
      "title": "もう夏の話",
    },
    Object {
      "date": 2020-03-08T15:22:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/32860?ima=0000&cd=member",
      "title": "なんでだろう",
    },
    Object {
      "date": 2020-03-01T12:00:00.000Z,
      "link": "http://www.keyakizaka46.com/s/k46o/diary/detail/32795?ima=0000&cd=member",
      "title": "TGCありがとうございました",
    },
  ],
  "json1": [Function],
  "options": Object {
    "copyright": "Seed & Flower LLC",
    "description": "「坂道シリーズ」第2弾　欅坂46",
    "feed": "http://www.keyakizaka46.com/",
    "feedLinks": Object {},
    "id": "http://www.keyakizaka46.com/",
    "link": "http://www.keyakizaka46.com/",
    "title": "欅坂46 藤吉 夏鈴 公式ブログ",
  },
  "rss2": [Function],
}
`;
